### PR TITLE
fix(slack): soften benign search no-result progress

### PR DESCRIPTION
**Pull request by kesslerio**
## What

Slack progress for optional local context searches no longer presents clean `rg` no-match results as scary exit-code failures. Direct `rg` exit 1 and `find ... | xargs rg ...` exit 123 now carry a narrow `benign_no_result` classification through exec command-output events, and Slack renders that as `No matches found` instead of `exit 123`.

The current head is merged with `origin/main` and removes the unrelated CI Node setup-helper changes from this PR. The effective diff is now limited to command outcome classification, agent progress metadata, and channel progress rendering.

## Why

The user-visible problem was noise: an agent could answer normally while Slack still showed a tool/progress warning for an optional markdown search that simply found nothing. Real failures remain visible because the classifier only downgrades known clean no-match shapes with empty output; missing paths, permission errors, command-not-found, syntax errors, timeouts, and unrelated nonzero exits still render as failures or exit-code statuses.

## Real behavior proof

**Behavior or issue addressed:** Slack-visible progress for a clean local `rg` no-match should say `No matches found`, not `exit 1`, `exit 123`, or failed.

**Real environment tested:** Local OpenClaw checkout on the feature branch after this patch, running real shell `rg` against the repository `docs` directory and then passing the real result through the production exec classifier and production Slack progress formatter.

**Exact steps or command run after this patch:** Ran a real no-match command with marker `__openclaw_no_such_marker_85478__` against `docs`, captured its real exit code/stdout/stderr, then invoked the production classifier and production channel streaming formatter via `node --import tsx`. The current head `1e24f832d68e5fe2149466bd0513063df813bbf4` merges current `origin/main`; the classifier/formatter behavior proven by the terminal proof is unchanged.

**Evidence after fix:** Console output from the after-fix run:

```json
{
  "rgExitCode": 1,
  "rgStdoutBytes": 0,
  "rgStderrBytes": 0,
  "outcomeClassification": "benign_no_result",
  "statusLabel": "No matches found",
  "slackProgress": "🛠️ No matches found"
}
```

**Observed result after fix:** The real clean no-match shell result is classified as `benign_no_result`, the Slack progress string is `🛠️ No matches found`, and the visible output contains no `exit 1`, `exit 123`, or `failed` wording.

**What was not tested:** Browser UI rendering was not tested because this change affects backend/tool event metadata and Slack progress rendering, not a browser route.

## Tests

- `pnpm test src/agents/bash-tools.exec-outcome-classification.test.ts src/agents/embedded-agent-subscribe.handlers.tools.test.ts src/plugin-sdk/channel-streaming.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `pnpm test src/auto-reply/reply/dispatch-from-config.test.ts src/auto-reply/reply/followup-runner.test.ts`
- `pnpm test src/auto-reply/reply/dispatch-from-config.test.ts -t "benign no-result progress"`
- `pnpm test src/auto-reply/reply/followup-runner.test.ts -t "benign no-result progress"`
- `pnpm test:bundled`
- `pnpm test src/agents/embedded-agent-runner/model.test.ts -t "builds an openai fallback for gpt-5.5 when the live catalog cache is cold"`
- `pnpm test src/auto-reply/reply/directive-handling.model.test.ts -t "reports Codex runtime auth for OpenAI status rows"`
- `pnpm exec oxfmt --check --threads=1 extensions/discord/src/monitor/message-handler.process.ts extensions/matrix/src/matrix/monitor/handler.ts extensions/msteams/src/reply-dispatcher.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts extensions/slack/src/monitor/message-handler/dispatch.ts extensions/telegram/src/bot-message-dispatch.ts src/agents/bash-tools.exec-outcome-classification.test.ts src/agents/bash-tools.exec-outcome-classification.ts src/agents/embedded-agent-subscribe.handlers.tools.test.ts src/agents/embedded-agent-subscribe.handlers.tools.ts src/auto-reply/get-reply-options.types.ts src/auto-reply/reply/agent-runner-execution.test.ts src/auto-reply/reply/agent-runner-execution.ts src/auto-reply/reply/followup-runner.test.ts src/auto-reply/reply/followup-runner.ts src/channels/streaming.ts src/infra/agent-events.ts src/infra/exec-outcome-classification-types.ts src/plugin-sdk/channel-streaming.test.ts`
- `git diff --check origin/main`

Result: all commands passed locally on head `1e24f832d68e5fe2149466bd0513063df813bbf4`. CI is running on the refreshed head. A later CI run hit a slow unrelated `src/agents/tools/image-tool.test.ts` case in `checks-node-agentic-agents`; that file is outside this PR diff, and the focused image-tool cases pass locally on both this PR branch and current `origin/main` (`2 passed`, about 195s on the PR branch and 133s on current main). The branch was merged with newer `origin/main` again, then the two long-running iMessage image-root tests were given a larger explicit timeout to stop the false timeout cascade observed in the required agents shard. The previously red `checks-fast-bundled-protocol` and `checks-node-agentic-agents` failures were stale-main failures; the local equivalents pass after merging current `origin/main`. The previous `checks-node-auto-reply-reply-dispatch` timeout also passes locally on the refreshed branch.

Browser test: skipped; this changes backend/tool event metadata and Slack progress rendering, not a browser route. `agent-browser` is installed.

Note: earlier versions of this PR included CI Node setup-helper hardening. That unrelated automation scope has been removed from the current head so this PR stays focused on benign command-result progress rendering.

## AI Assistance

Implemented with Codex using the repository test wrappers above. Review pass found no residual actionable work.

### CODE REVIEW

**Comment by social5h3ll** [comment]
Hey @kesslerio / @clawsweeper hatch - I've implemented an improved version of this feature at social5h3ll/openclaw#1 that fixes a NULL dereference bug in the original implementation and keeps the classifier strictly narrow (only rg exit 1 and xargs-rg exit 123 patterns qualify as benign).

The key improvement: readExecCommandFromArgs was called with startData?.args when startData can be undefined, causing crashes in tests and edge cases. Fixed by making the parameter optional.

All 110 tests pass across the 4 relevant test suites. Happy to close this PR and fold the changes in if preferred, or keep as a reference implementation.

**Comment by social5h3ll** [comment]
Heads up — I've moved the implementation to a fresh PR with the narrow ripgrep-only classifier: #85498.

Key fixes from our earlier closed attempt:
- Classifier is now conservative: only direct `rg` exit 1 and `xargs`-wrapped `rg` exit 123 qualify as `benign_no_result`
- No broader exit-code-1 softening
- PR is rebased on latest main and mergeable

File: src/agents/bash-tools.exec-outcome-classification.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+const EXIT_FOOTER_RE = /\n*\(Command exited with code \d+\)\s*$/u;
+const RG_TOKEN_RE = /(?:^|[\s|;&()])rg(?:\s|$)/iu;
+const XARGS_TOKEN_RE = /(?:^|[\s|;&()])xargs(?:\s|$)/iu;
+
+function cleanAggregateOutput(value: string | undefined): string {
+  return (value ?? "").replace(EXIT_FOOTER_RE, "").trim();
+}
+
+function isDirectRgNoMatch(params: { command: string; exitCode: number }): boolean {
+  return params.exitCode === 1 && RG_TOKEN_RE.test(params.command);
```
[the remark this answers is not in the document]
Comment (kesslerio) [reply src/agents/bash-tools.exec-outcome-classification.ts]: Fixed in 82ffad9848196ade642984da3f7b187dfc7dc8dd by replacing the token regex with command-shape parsing. The classifier now only treats direct rg exit 1 or final-segment xargs rg exit 123 with empty aggregate output as benign, rejects separators/downstream failures like `echo rg >/dev/null; false` and `xargs rg | xargs false`, and includes regression coverage in `src/agents/bash-tools.exec-outcome-classification.test.ts`. Verified with focused tests, lint, docs, changed checks, and a clean local Codex review.

**Comment by kesslerio** [comment]
Follow-up is done.

What changed after review:

- Tightened the benign `rg` classification. It now only marks real no-match searches as benign: direct `rg` exit 1, or final-segment `xargs rg` exit 123, with no aggregate output.
- Added regression coverage for the bad cases the bot called out, including `echo rg >/dev/null; false`, shell separators, newline-separated commands, background commands, and downstream pipeline failures like `xargs rg | xargs false`.
- Carried the new `outcomeClassification` / `statusLabel` fields through the embedded tool event path and the channel streaming path.
- Updated Slack, Discord, Telegram, Matrix, and Teams progress rendering so benign no-result searches show `No matches found` instead of scary exit-code text.
- Fixed two unrelated CI blockers that showed up while this branch was running: the BWS resolver now preserves parse causes, and the docs redirect for `/reference/logs` points to `/logging`.

Verification run locally:

- Focused Vitest coverage for classifier, embedded tool events, auto-reply runners, channel streaming, and Slack fallback rendering.
- `pnpm lint --threads=8`
- `pnpm check:docs`
- `pnpm check:changed`
- Local `codex review --base origin/main` after the final patch: no actionable correctness issues found.

CI is green on the latest head, `82ffad9848196ade642984da3f7b187dfc7dc8dd`.

**Comment by rteoo** [comment]
Tested this bug class against a local stable field case, with private paths/session details redacted.

We saw a Discord-visible warning for a diagnostic search that exited `1` with empty output. The command was an `rg` search for prior status-label text in local session/log files; no matches were found, and the agent continued normally. There was no runtime/gateway/cron failure.

So this PR's direction looks correct for at least the `rg` no-match case: a clean no-match should be rendered as “no matches found” / benign progress, not as a failed tool alert.

Adjacent note: we also saw a separate `rg` exit `2` from using lookahead without `--pcre2`; that one should remain a real command error. The useful distinction is `rg` exit `1` with no output vs parser/runtime errors like exit `2`.

**Comment by kesslerio** [comment]
@clawsweeper re-review

Addressed the remaining P2 docs-tree finding in `93a45fa352`.

I removed the two tracked scratch/planning docs from the public docs tree:

- `docs/brainstorms/2026-05-22-slack-benign-tool-failure-ux-requirements.md`
- `docs/plans/2026-05-22-slack-benign-tool-failure-ux.md`

I left the classifier, channel event plumbing, progress rendering, tests, and the `/reference/logs` docs redirect unchanged.

Verification:

- `git diff --check`
- `node scripts/run-vitest.mjs src/agents/bash-tools.exec-outcome-classification.test.ts src/plugin-sdk/channel-streaming.test.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `pnpm check:docs`

**Comment by kesslerio** [comment]
Follow-up after the ClawSweeper pass:

- Removed the two public scratch docs ClawSweeper flagged.
- Merged latest `origin/main` so the PR picks up the current guard fixes.
- Fixed two current-head guard nits that showed up after the merge: one unnecessary test cast in the Codex app-server test, and the Feishu doctor import now uses the focused plugin SDK config-contracts subpath.

Local verification after those changes:

- `git diff --check`
- `pnpm lint --threads=8`
- `pnpm check:architecture`
- `pnpm check:docs`
- `node scripts/run-vitest.mjs src/agents/bash-tools.exec-outcome-classification.test.ts src/plugin-sdk/channel-streaming.test.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `node scripts/run-vitest.mjs extensions/codex/src/app-server/run-attempt.test.ts -t "ignores turn/completed notifications for other subscribed threads"`

ClawSweeper has moved the PR to `status: ready for maintainer look` on head `b946ce74747f6d8650e7adf102d1ad87a1acb311`.

**Comment by kesslerio** [comment]
@openclaw-mantis telegram desktop proof

Please verify a clean `rg` no-match command progress renders as `No matches found`, and an `rg` syntax/runtime error still shows failure wording.

SDK compatibility note for maintainers:

This PR intentionally adds two optional command-output progress metadata fields: `outcomeClassification` and `statusLabel`.

Compatibility properties:

- Both fields are optional.
- Unset behavior is unchanged.
- Existing channel renderers still fall back to exit/status rendering.
- The classifier is narrow: only direct `rg` exit 1 or final-segment `xargs rg` exit 123 with empty aggregate output can become `benign_no_result`.
- Parser errors, permission errors, command-not-found, timeouts, non-empty output, shell separators, and downstream pipeline failures remain normal failures.

This is intended as additive SDK progress metadata, not a replacement for command exit/status semantics.

**Comment by kesslerio** [comment]
CI note on the remaining red check:

`checks-windows-node-test` failed during setup before project tests ran:

```text
Expected Node '24.x', but active node is '22.19.0' at /c/Program Files/nodejs/node
```

I’m treating this as Windows runner/setup drift unless maintainers want a branch-side CI adjustment. The branch-specific local checks and focused test suites listed above passed, and the Mantis Telegram desktop proof request is queued separately for visible transport evidence.

**Comment by kesslerio** [comment]
Follow-up for the remaining Windows CI failure:

I cherry-picked the current-main CI helper fix `c84f61cd2e` onto this branch as `d77958fb79`.

Why: `checks-windows-node-test` was failing before tests ran because Git Bash kept resolving `node` to `C:/Program Files/nodejs/node` 22.19.0 after requesting `24.x`. The helper now normalizes Windows toolcache paths before prepending PATH and writes the Windows path to `GITHUB_PATH`.

Verification:

- `git diff --check HEAD~1..HEAD`
- `pnpm exec oxfmt --check --threads=1 .github/actions/setup-pnpm-store-cache/ensure-node.sh test/scripts/setup-pnpm-store-cache-ensure-node.test.ts`
- Manual shell harness confirmed a Windows-style toolcache path resolves to Node 24.15.0 and writes `C:\hostedtoolcache\windows\node\24.15.0\x64` to `GITHUB_PATH`.

Note: the focused Vitest file was not useful locally; it hung in this checkout's transform path before producing test output, so I used the direct shell harness for the behavior the Windows job needs.

**Comment by kesslerio** [comment]
@clawsweeper re-review

Follow-up after Oracle's second-opinion pass on the rank/readiness gap.

Oracle's read matched the durable ClawSweeper concern: the remaining blocker was not the classifier itself, but the full visible progress sequence. A terminal command item could still carry raw no-result footer text before the sibling `command_output` event rendered `No matches found`.

What changed on the latest head `3b9665554014157560adfec147e0c9e2b21b9392`:

- `49a70153a7` rewrites the terminal command item summary to `No matches found` when the exec outcome is `benign_no_result`.
- `49a70153a7` adds a handler-level regression asserting the item event summary no longer contains `(Command exited with code 123)` while the sibling `command_output` event still preserves `exitCode: 123`, `outcomeClassification: benign_no_result`, and `statusLabel: No matches found`.
- `3b96655540` extends the Slack regression so item-event and command-output event arrive in the same progress stream, which is the exact sequence ClawSweeper called out.
- `01d755dcca` hardens the Windows Node fallback per Oracle's CI review: wildcard download resolution now fails cleanly instead of producing an empty version, and PowerShell zip extraction receives paths via env vars so apostrophes in `RUNNER_TEMP` cannot break parsing.

Deterministic proof run after the final patch:

```json
{
  "itemSummary": "No matches found",
  "commandOutcomeClassification": "benign_no_result",
  "commandStatusLabel": "No matches found",
  "commandExitCode": 123,
  "visibleLines": [
    "🛠️ find files named \"*.md\" in docs -> run xargs rg",
    "🛠️ No matches found; command find files named \"*.md\" in docs -> run xargs rg"
  ],
  "containsRawExitInVisibleLines": false
}
```

Verification run locally:

- `bash -n .github/actions/setup-pnpm-store-cache/ensure-node.sh`
- `git diff --check -- .github/actions/setup-pnpm-store-cache/ensure-node.sh test/scripts/setup-pnpm-store-cache-ensure-node.test.ts`
- `git diff --check -- src/agents/pi-embedded-subscribe.handlers.tools.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `pnpm exec oxfmt --check --threads=1 test/scripts/setup-pnpm-store-cache-ensure-node.test.ts`
- `pnpm exec oxfmt --check --threads=1 src/agents/pi-embedded-subscribe.handlers.tools.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- Direct Windows helper harness: Windows-style `RUNNER_TOOL_CACHE` resolves to Node 24.15.0.
- Direct Windows zip fallback harness: `RUNNER_TEMP` containing an apostrophe resolves, env-var PowerShell paths are used, and Node 24.15.0 activates.
- Direct handler/channel proof: the item event plus command-output event produce no raw exit-code wording in visible progress lines.

Note: focused Vitest commands in this local checkout timed out in the repo's Rolldown transform path before producing test-body output, which is the same local runner issue noted earlier. I used direct harnesses for the behavior under review and left CI to validate the repository test runners.

**Comment by kesslerio** [comment]
Proof addendum for the ClawSweeper rank-up checklist, latest head `9942a9955141f348a07ddab60a7d25c365b214f1`.

I added the paired real-error Slack regression in `9942a99551` so the proof now covers both sides:

- clean `rg` no-match -> `No matches found`
- real `rg` error -> visible `exit 2`, not `No matches found`

Direct proof transcript from the shared channel progress formatter:

```json
{
  "benignVisibleLines": [
    "🛠️ No matches found",
    "🛠️ No matches found; command find docs -name '*.md' -print0 | xargs -0 rg 'missing phrase'"
  ],
  "benignContainsRawExit": false,
  "errorVisibleLines": [
    "🛠️ rg: missing-path: No such file or directory (Command exited with code 2)",
    "🛠️ exit 2; command rg 'needle' missing-path"
  ],
  "errorContainsExit2": true,
  "errorContainsNoMatchesFound": false
}
```

Additional verification after this addendum:

- `pnpm exec oxfmt --check --threads=1 extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `git diff --check -- extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- Direct formatter proof above.

This completes source fix, full item+command Slack sequence regression, paired real-error regression/proof, and proof comment on the final head.

**Comment by kesslerio** [comment]
@clawsweeper re-review

Final addendum pushed in `9942a99551`: added the paired Slack real-error regression and proof comment so clean no-match and real `rg` error behavior are both covered on the latest head.

**Comment by kesslerio** [comment]
Follow-up pushed for the latest CI artifact failure.

Head is now `7e89f4fbf3e2d736a56a0582dc79e71a15a9397a`.

What changed:
- The Windows Node resolver-failure regression now requests `99.x` instead of `24.x`, so CI cannot accidentally satisfy the test from a real hosted Node 24 toolcache.
- This keeps the intended assertion: failed wildcard resolution must not produce an empty `node--...` / `node-v-win...` download path.

Verified locally:

```sh
node scripts/run-vitest.mjs run --config test/vitest/vitest.tooling.config.ts test/scripts/setup-pnpm-store-cache-ensure-node.test.ts -t "fails wildcard download"
node scripts/run-vitest.mjs run --config test/vitest/vitest.tooling.config.ts test/scripts/setup-pnpm-store-cache-ensure-node.test.ts
```

@clawsweeper re-review after this CI fix, please.

**Comment by kesslerio** [comment]
CI note: the remaining `check-guards` failure appears to be current-main shrinkwrap drift, not this PR's Slack/Node patch.

The failing guard is:

```text
pnpm deps:shrinkwrap:check
generated npm-shrinkwrap.json contains package versions absent from pnpm-lock.yaml:
node_modules/@aws-sdk/core locked @aws-sdk/core@3.974.14; ...
```

I reproduced the same failure on a clean detached `origin/main` worktree at `b36fa1d8f1` with:

```sh
node scripts/generate-npm-shrinkwrap.mjs --all --check
```

So I am not folding a dependency/shrinkwrap refresh into this already-large Slack PR unless maintainers want that combined.

**Comment by kesslerio** [comment]
Refreshed this branch against current `origin/main` to clear the merge conflicts.\n\nHead now: `9f34dfbde3e27a38569e6d2d65231227b9254042`\n\nConflict resolution notes:\n\n- kept the hardened Windows Node fallback from this PR while preserving main's additional Windows arch matching\n- kept main's CLI fallback rollback path in `agent-runner-execution.ts`; the PR's embedded item/command progress handling remains in the embedded event path\n- adjusted the real-error Slack regression to assert the current visible error footer, `Command exited with code 2`, while still proving it is not classified/rendered as `No matches found`\n\nValidation run locally:\n\n```sh\nbash -n .github/actions/setup-pnpm-store-cache/ensure-node.sh\n/tmp/openclaw-pr76920-merge/node_modules/.bin/oxfmt --check --threads=1 \\n  src/auto-reply/reply/agent-runner-execution.ts \\n  test/scripts/setup-pnpm-store-cache-ensure-node.test.ts \\n  extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts \\n  src/agents/pi-embedded-subscribe.handlers.tools.test.ts\nnode scripts/test-projects.mjs \\n  test/scripts/setup-pnpm-store-cache-ensure-node.test.ts \\n  extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts \\n  src/agents/pi-embedded-subscribe.handlers.tools.test.ts\n```\n\nResults: shell syntax passed, formatting passed, and focused Vitest passed 3 shards / 95 tests.\n\n@clawsweeper re-review

**Comment by kesslerio** [comment]
Follow-up for the latest CI failures on head `9f34dfbde3`.

Head is now `1d6f638a9324bfd677d0a12bf98200ec6c16106e`.

Root cause:

- `check-prod-types` was correct: `readCommandOutputClassification` became unused after the conflict-resolution merge.
- `checks-node-auto-reply-reply-agent-runner` was also correct: the main `runAgentTurnWithFallback` command-output forwarding path had dropped `outcomeClassification` and `statusLabel` again.

Patch:

- Forward `outcomeClassification` and `statusLabel` from embedded `command_output` events through `runAgentTurnWithFallback`.

Verification run locally:

```sh
/tmp/openclaw-pr76920-merge/node_modules/.bin/oxfmt --check --threads=1 \
  src/auto-reply/reply/agent-runner-execution.ts \
  src/auto-reply/reply/agent-runner-execution.test.ts

node scripts/test-projects.mjs src/auto-reply/reply/agent-runner-execution.test.ts

node scripts/run-tsgo.mjs -p tsconfig.core.json --incremental --tsBuildInfoFile .artifacts/tsgo-cache/core.tsbuildinfo
node scripts/run-tsgo.mjs -p tsconfig.extensions.json --incremental --tsBuildInfoFile .artifacts/tsgo-cache/extensions.tsbuildinfo
```

Results: formatting passed; the failing agent-runner shard passed 137/137; direct prod type checks passed.

`pnpm tsgo:prod` was not usable in this temp worktree because pnpm tried to purge the symlinked `node_modules` without a TTY, so I ran the underlying CI commands directly.

@clawsweeper re-review

**Comment by kesslerio** [comment]
Current-head follow-up after the last patch:

- Head `1d6f638a9324bfd677d0a12bf98200ec6c16106e` now has green CI, including the previously failing `check-prod-types` and `checks-node-auto-reply-reply-agent-runner` jobs.
- The Windows Node helper check is also green.
- The older Codex inline thread about overbroad `rg` no-match classification has been resolved; that concern was fixed earlier by the narrow command-shape classifier and regression tests.

No remaining failing checks or unresolved review threads are visible from my side. Waiting on ClawSweeper label/review sync.

**Comment by kesslerio** [comment]
Rebased this PR onto current `origin/main` and resolved the merge conflicts.

Changes in this refresh:
- Kept the new shared channel streaming location (`src/channels/streaming.ts`) and left `src/plugin-sdk/channel-streaming.ts` as the compatibility shim.
- Carried the benign `rg` no-result classification through command-output progress for Slack, Discord, Matrix, MS Teams, and Telegram.
- Applied the source-side fix so terminal command item summaries render `No matches found` for benign no-result commands instead of leaking the raw `(Command exited with code 123)` footer.
- Added/kept the full Slack regression where the same progress draft receives both the command item event and the command-output event.
- Hardened the Windows Node fallback helper while rebasing, including native Windows toolcache path normalization, non-empty Node download version resolution, and env-var based `Expand-Archive` invocation.

Local verification on head `c157ff29db56e98da6b490cc2dce9d54fa84b661`:

```sh
pnpm test \
  src/agents/bash-tools.exec-outcome-classification.test.ts \
  src/agents/embedded-agent-subscribe.handlers.tools.test.ts \
  src/plugin-sdk/channel-streaming.test.ts \
  extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts \
  test/scripts/setup-pnpm-store-cache-ensure-node.test.ts

pnpm check:test-types
pnpm lint --threads=8
pnpm exec oxfmt --check --threads=1 \
  .github/actions/setup-pnpm-store-cache/ensure-node.sh \
  extensions/discord/src/monitor/message-handler.process.ts \
  extensions/matrix/src/matrix/monitor/handler.ts \
  extensions/msteams/src/reply-dispatcher.ts \
  extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts \
  extensions/slack/src/monitor/message-handler/dispatch.ts \
  extensions/telegram/src/bot-message-dispatch.ts \
  src/agents/bash-tools.exec-outcome-classification.test.ts \
  src/agents/bash-tools.exec-outcome-classification.ts \
  src/agents/embedded-agent-subscribe.handlers.tools.test.ts \
  src/agents/embedded-agent-subscribe.handlers.tools.ts \
  src/auto-reply/get-reply-options.types.ts \
  src/auto-reply/reply/agent-runner-execution.ts \
  src/channels/streaming.ts \
  src/plugin-sdk/channel-streaming.test.ts \
  test/scripts/setup-pnpm-store-cache-ensure-node.test.ts
```

All passed locally. CI is rerunning on the refreshed head now.

@clawsweeper re-review

**Comment by kesslerio** [comment]
@clawsweeper re-review

Current head `1e24f832d68e5fe2149466bd0513063df813bbf4` is green in CI. I also removed the unrelated CI Node setup-helper scope from this PR, leaving the diff focused on benign command-result classification/progress rendering. The only extra non-feature change is the timeout bump for two existing long-running iMessage image-root tests that were repeatedly timing out in the required agents shard; focused local proof is noted in the PR body.

**Comment by kesslerio** [comment]
@clawsweeper re-review

Addressed the two P2 findings from the latest review on head `e122fbe94ba865cf8d32039c136ec2732e250088`:

- direct reply suppression now ignores `outcomeClassification: "benign_no_result"` command outputs even when exitCode is 1/123
- queued follow-up suppression now does the same
- added focused direct-reply and queued-follow-up regressions proving benign no-result progress does not suppress later real tool-error warnings

Validation run locally:
- `pnpm test src/auto-reply/reply/dispatch-from-config.test.ts -t "benign no-result progress"`
- `pnpm test src/auto-reply/reply/followup-runner.test.ts -t "benign no-result progress"`
- `pnpm test src/auto-reply/reply/dispatch-from-config.test.ts src/auto-reply/reply/followup-runner.test.ts`
- `pnpm test src/agents/bash-tools.exec-outcome-classification.test.ts src/agents/embedded-agent-subscribe.handlers.tools.test.ts src/plugin-sdk/channel-streaming.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts`
- `pnpm exec oxfmt --check --threads=1 src/auto-reply/reply/dispatch-from-config.ts src/auto-reply/reply/dispatch-from-config.test.ts src/auto-reply/reply/followup-runner.ts src/auto-reply/reply/followup-runner.test.ts`
- `git diff --check origin/main`

**Comment by kesslerio** [comment]
CI refresh note for head `772c4a68cfc75ac80171080ec3c39ad29bb4fd69`:

- merged current `origin/main` into the PR branch after main advanced
- refreshed the generated diffs viewer runtime asset so `pnpm build:ci-artifacts` leaves the tree clean
- reproduced the prior red `build-artifacts` failure mode locally: gateway-watch failed when the artifact build left `extensions/diffs/assets/viewer-runtime.js` dirty
- verified locally after the generated asset refresh:
  - `pnpm build:ci-artifacts`
  - `node scripts/check-gateway-watch-regression.mjs --skip-build --ready-timeout-ms 5000`
  - `pnpm test src/auto-reply/reply/dispatch-from-config.test.ts src/auto-reply/reply/followup-runner.test.ts`
  - `pnpm test src/agents/bash-tools.exec-outcome-classification.test.ts src/agents/embedded-agent-subscribe.handlers.tools.test.ts src/plugin-sdk/channel-streaming.test.ts extensions/slack/src/monitor/message-handler/dispatch.preview-fallback.test.ts -t "benign no-result|rg no-match|command item|command-output|No matches"`
  - `pnpm exec oxfmt --check --threads=1 src/auto-reply/reply/dispatch-from-config.ts src/auto-reply/reply/dispatch-from-config.test.ts src/auto-reply/reply/followup-runner.ts src/auto-reply/reply/followup-runner.test.ts`
  - `git diff --check origin/main`

CI is running again now on the refreshed head.

**Comment by BingqingLyu** [comment]
### :warning: Cross-PR Conflict Detected

This PR shares modified code with #85200, #85235, #85381, #85403, #85465, #85507, #85612, #85904, #85992, #86624, #86634, #86672, #86950, #87072, #87171, #87200, #87339, #87433, #87463.

**Shared functions:**

| Function | File | Also modified by |
|----------|------|-----------------|
| `createFollowupRunner` | `src/auto-reply/reply/followup-runner.ts` | #85381, #85904, #86624, #86950, #87463 |
| `createMatrixRoomMessageHandler` | `extensions/matrix/src/matrix/monitor/handler.ts` | #87200 |
| `dispatchPreparedSlackMessage` | `extensions/slack/src/monitor/message-handler/dispatch.ts` | #85507, #85612 |
| `dispatchReplyFromConfig` | `src/auto-reply/reply/dispatch-from-config.ts` | #85235, #85403, #85904, #85992, #86634, #86672, #87171 |
| `dispatchTwiceWithFreshDispatchers` | `src/auto-reply/reply/dispatch-from-config.test.ts` | #85235, #85904, #85992, #86634, #86672, #87171, #87200 |
| `processDiscordMessage` | `extensions/discord/src/monitor/message-handler.process.ts` | #85200, #85465, #87339, #87433 |
| `resolveTelegramReasoningLevel` | `extensions/telegram/src/bot-message-dispatch.ts` | #85403, #87072 |

**Recommendation:** 20 PRs share 7 shared functions across 5 areas. Coordinate with #85200, #85235, #85381 and 16 others before merging.

---
*PR conflict auto-detected by [codegraph](https://github.com/alibaba/neug/tree/main/skills/codegraph) — a code graph analysis tool built on [neug](https://github.com/alibaba/neug).*

**Comment by kesslerio** [comment]
Closing this PR. During reassessment (2026-07-02), the branch is 10,709 commits behind `main` and CI (`build-artifacts`) is failing. More fundamentally, `main` has re-architected the progress/draft rendering path this PR hooks into: the `outcomeClassification`/`statusLabel` fields this PR adds to the Slack dispatch and `channel-streaming` have been replaced by the `ChannelProgressDraftLine` / `resolveChannelProgressDraftConfig` system (`buildChannelProgressDraftLine`, `createChannelProgressDraftGate`, `mergeChannelProgressDraftLine`). The files this PR modifies (`pi-embedded-subscribe.handlers.tools.ts`, `bash-tools.exec-outcome-classification.ts`, `exec-outcome-classification-types.ts`) were renamed or removed on main. The feature intent (soften benign search no-result progress) is still not shipped on main, but landing it now requires a rewrite against the new `ChannelProgressDraftLine` API rather than a rebase. I'll file a fresh PR if I reimplement against the new system. Thanks to the reviewers who looked at the earlier iteration.
